### ISSUE TITLE
CoreML export of YOLOv11 model produces incorrect or missing bounding boxes, while TFLite works as expected

### ISSUE THREAD

**Issue by jandrop** [ultralytics/ultralytics#22309]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

I trained a YOLOv11 model using the same dataset and training parameters. When exporting the trained model to both TFLite and CoreML formats, I observed the following issue:

The TFLite model returns correct detection results.
The CoreML model either returns no detections or produces bounding boxes that are missing or located outside the image boundaries.

I've attached the example notebook that I'm using. And two images of the output as a reference.

![Image](https://github.com/user-attachments/assets/d914e824-8bd8-4980-b216-6add6be4f905)
![Image](https://github.com/user-attachments/assets/98872a8f-edc5-418c-b586-dc3555577d84)

[Train_YOLO_Models.ipynb](https://github.com/user-attachments/files/22751535/Train_YOLO_Models.ipynb)

### Environment

```
Ultralytics 8.3.206 🚀 Python-3.11.11 torch-2.8.0 CPU (Apple M4 Pro)
Setup complete ✅ (14 CPUs, 48.0 GB RAM, 758.9/926.4 GB disk)

OS                     macOS-26.0.1-arm64-arm-64bit
Environment            Darwin
Python                 3.11.11
Install                pip
Path                   /Users/jandrop/Developer/AI/Yolo/.venv/lib/python3.11/site-packages/ultralytics
RAM                    48.00 GB
Disk                   758.9/926.4 GB
CPU                    Apple M4 Pro
CPU count              14
GPU                    None
GPU count              None
CUDA                   None

numpy                  ✅ 2.1.3>=1.23.0
matplotlib             ✅ 3.10.6>=3.3.0
opencv-python          ✅ 4.12.0.88>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.2>=1.4.1
torch                  ✅ 2.8.0>=1.8.0
torch                  ✅ 2.8.0!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.23.0>=0.9.0
psutil                 ✅ 7.1.0
polars                 ✅ 1.34.0
ultralytics-thop       ✅ 2.0.17>=2.0.0
```

### Minimal Reproducible Example

The notebook with the reproducible code is attached

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jandrop, thank you for your interest in Ultralytics 🚀! This is an automated response — an Ultralytics engineer will review and assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For CoreML export/detections issues specifically, please include:
- Exact export command(s) used (including args like imgsz, nms, iou, conf, half, int8)
- Model variant (e.g., YOLO11n/s/m/l/x) and a link to a tiny sample of the trained weights if possible
- coremltools version, macOS/iOS version, and how you’re running inference (Core ML Tools Python, Xcode preview, Vision framework, etc.)
- A single example input image and the raw CoreML output you receive vs. expected results (TFLite output you referenced)
- Whether you enabled built-in NMS on export and/or applied any custom post-processing downstream

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [Python>=3.8](https://www.python.org/) environment with [PyTorch>=1.8](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Additional notes for your report:
- Thanks for sharing the notebook and screenshots — if possible, please reduce the notebook to the smallest snippet that reproduces CoreML misaligned/missing boxes and attach a single small test image to run end-to-end
- Please include the exact CoreML export options you used and confirm whether TFLite and PyTorch inference on the same image produce the expected boxes with the same thresholds
- If you have a small CoreML model file that reproduces the issue, sharing it would help us reproduce locally

🙏 Once you provide the above details, it will greatly speed up triage. An Ultralytics engineer will follow up shortly.

**Comment by glenn-jocher** [comment]
@jandrop CoreML export with NMS and Xcode Preview in macOS 26 works correctly for me with:

```bash
yolo export format=coreml nms
```

Can you produce code to reproduce the bug you are seeing?

<img width="964" height="1055" alt="Image" src="https://github.com/user-attachments/assets/23c86e21-4635-445b-869e-1e8fc8513903" />

**Comment by jandrop** [comment]
> [@jandrop](https://github.com/jandrop) CoreML export with NMS and Xcode Preview in macOS 26 works correctly for me with:
> 
> yolo export format=coreml nms
> Can you produce code to reproduce the bug you are seeing?
> 
> <img alt="Image" width="964" height="1055" src="https://private-user-images.githubusercontent.com/26833433/498481134-23c86e21-4635-445b-869e-1e8fc8513903.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTk4NjQzMjEsIm5iZiI6MTc1OTg2NDAyMSwicGF0aCI6Ii8yNjgzMzQzMy80OTg0ODExMzQtMjNjODZlMjEtNDYzNS00NDViLTg2OWUtMWU4ZmM4NTEzOTAzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDA3VDE5MDcwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTEzYWMxOWEyMTkzMmNhMDE1ZDk0YjAxMjg5MDg0ZjIyNDM0ZTE1YjQ0ZmZiMmEwZTc4NTMwZmQzMTE3YWQwZDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iFVaiPLdiIFdMtjQXKhEQChBvfhDsTzLyxbj6dqPUYk">

Hi, thanks for your help, I've attached in the issue the complete notebook to create and train the model. https://github.com/user-attachments/files/22751535/Train_YOLO_Models.ipynb

I use this code to create the mlpackage

```
# Export to CoreML
from ultralytics import YOLO
model = YOLO('runs/detect/train/weights/best.pt')

model.export(
    format='coreml',
    nms=True
)
```
[EDIT]
[best.pt.zip](https://github.com/user-attachments/files/22752504/best.pt.zip)

This is the model that I use to export to mlpackage

**Comment by Y-T-G** [comment]
@jandrop Thanks for the model. I can reproduce it with the model. It's a strange behavior by CoreML. Exporting with NMS works with your model if you export with `format='mlmodel'`. But `format='coreml'` doesn't work. I did some debugging and the behavior is related to number of classes in the dataset. For Ultralytics MS COCO pretrained models, exporting as both `mlpackage` and `mlmodel` works with NMS. But if the number of classes is not a multiple of 80, it doesn't work, which is the case with your model.

We could add a workaround by manually padding the classes to multiple of 80.

**Comment by Y-T-G** [comment]
You can try exporting with

```
pip install git+https://github.com/ultralytics/ultralytics@coreml-nms-custom
```

**Comment by jandrop** [comment]
> You can try exporting with
> 
> ```
> pip install git+https://github.com/ultralytics/ultralytics@coreml-nms-custom
> ```

Hi, Thanks for the workaround, I will try it and put the results editing this message.

[EDIT]:
When I export to CoreML I got 80 more classes, not a multiple of 80

<img width="1561" height="976" alt="Image" src="https://github.com/user-attachments/assets/3c1505d6-cd70-47da-ad4b-801ef7f98149" />

but it seems the inference works on the preview:

<img width="1561" height="976" alt="Image" src="https://github.com/user-attachments/assets/d71d3f37-e71f-4069-aeec-980f7b416698" />

Now the problem is on iOS when we use this model, it works worse because it has more false positive detections with a 99% of confidence. No idea if it's because of the iOS model or, in this case, the iOS library for Ultralytics on Flutter.

**Comment by jandrop** [comment]
I modified the exported with this to add the exact multiple of 80, I'm not a python expert but it seems to work.

```
if self.coreml and self.nc % 80 != 0:
    remainder = self.nc % 80
    pad_length = 80 - remainder
    cls = torch.nn.functional.pad(cls, (0, pad_length, 0, 0), "constant", 0)
```

**Comment by Y-T-G** [comment]
> When I export to CoreML I got 80 more classes, not a multiple of 80

Ah yes, you're right. Forgot to add the subtraction.

**Comment by Y-T-G** [comment]
Can you try exporting using the updated PR again and report whether it works on iOS? 

`pip install git+https://github.com/ultralytics/ultralytics@coreml-nms-custom`

There were several reports before that they were getting high confidence (on iOS specifically, the preview confidence issue is a different bug). But I couldn't reproduce it with Ultralytics. If this also resolves that issue somehow, then that's great.

https://github.com/ultralytics/ultralytics/issues/13079

**Comment by glenn-jocher** [comment]
@Y-T-G oh I see, this is exported using an older format from Apple. I think we should probably just drop support for the older coreml format, it's quite old now and Apple's not returning to it.

**Comment by glenn-jocher** [comment]
Here is the line that still allows the older format that we should drop support for (was introduced in 2023 to ease the transition):
https://github.com/ultralytics/ultralytics/blob/f7e89e772ac5aadd6c64dd43bf190d675d9edf57/ultralytics/engine/exporter.py#L872

**Comment by Y-T-G** [comment]
@glenn-jocher The bug is unfortunately with the newer format. The older one `mlmodel` works fine. The newer `mlpackage` one has the bug.

**Comment by glenn-jocher** [comment]
@Y-T-G maybe the bug was accidentally introduced in the recent NMS code changes for CoreML? Can you try to reproduce from before that PR?

**Comment by Y-T-G** [comment]
@glenn-jocher I tried Ultralytics 8.3.195 now, before the cleanup PR:
https://github.com/ultralytics/ultralytics/pull/22310

It still has the bug.

There are older scattered reports of this issue, i.e. `format=coreml` with `nms=True` not working with their model but working with Ultralytics pretrained models:
https://github.com/ultralytics/ultralytics/issues/20061#issuecomment-2798858878
https://github.com/ultralytics/ultralytics/issues/19857#issuecomment-2802285761

**Comment by glenn-jocher** [comment]
Ah ok! This is strange then, there must be some part of the CoreML or CoreML NMS export that is hard-coded to COCO models. I'm trying to reproduce too with a VOC-trained model.

**Comment by Y-T-G** [comment]
Yes. I also noticed the number of classes affect the final confidence, which it shouldn't. CoreML NMS does something behind the scenes. I think the NMS pipeline is an internal API that Apple does not expect us to use. There's little to no documentation on it.

https://github.com/apple/coremltools/issues/1773#issuecomment-1433755513

**Comment by alejandroplatas** [comment]
> Can you try exporting using the updated PR again and report whether it works on iOS?
> 
> `pip install git+https://github.com/ultralytics/ultralytics@coreml-nms-custom`
> 
> There were several reports before that they were getting high confidence (on iOS specifically, the preview confidence issue is a different bug). But I couldn't reproduce it with Ultralytics. If this also resolves that issue somehow, then that's great.
> 
> [#13079](https://github.com/ultralytics/ultralytics/issues/13079)

Hi, sorry for the late response, using this, the number of classes now are multiple of 80.

**Comment by jandrop** [comment]
> Can you try exporting using the updated PR again and report whether it works on iOS?
> 
> `pip install git+https://github.com/ultralytics/ultralytics@coreml-nms-custom`
> 
> There were several reports before that they were getting high confidence (on iOS specifically, the preview confidence issue is a different bug). But I couldn't reproduce it with Ultralytics. If this also resolves that issue somehow, then that's great.
> 
> [#13079](https://github.com/ultralytics/ultralytics/issues/13079)

Hi, sorry for the late response, using this, now the number of classes are now multiple of 80.

**Comment by jandrop** [comment]
> Yes. I also noticed the number of classes affect the final confidence, which it shouldn't. CoreML NMS does something behind the scenes. I think the NMS pipeline is an internal API that Apple does not expect us to use. There's little to no documentation on it.
> 
> [apple/coremltools#1773 (comment)](https://github.com/apple/coremltools/issues/1773#issuecomment-1433755513)

I notice the same thing. If you look at both pictures, the left one is the model created during training in the .pt file, and the right one is the model exported as .coreml. The confidence is different. Also in the third picture the right one has one more object than the left one 

<img width="1136" height="974" alt="Image" src="https://github.com/user-attachments/assets/9000fb9f-8c6a-4e07-840c-25f868706dea" />

**Comment by Y-T-G** [comment]
@jandrop Can you try comparing confidence while running the PyTorch model predictions with `rect=False`

**Comment by jandrop** [comment]
> [@jandrop](https://github.com/jandrop) Can you try comparing confidence while running the PyTorch model predictions with `rect=False`

@Y-T-G  This is the inference with rect=False.

The confidence is still different, but now the third picture has the same objects number.

<img width="1421" height="1265" alt="Image" src="https://github.com/user-attachments/assets/e1456507-74ce-42c6-a1fb-116765c0aec5" />

**Comment by glenn-jocher** [comment]
I think let's merge this in a new release to at least resolve the main issue of no-detections, then @Y-T-G we should also raise an issue on the coremltools repo.

**Comment by UltralyticsAssistant** [comment]
Great news! A potential fix has been merged in the pull request ultralytics 8.3.207 Apple MLProgram NMS models 80-class pad workaround. This aims to resolve CoreML export NMS issues for detection models on iOS, especially when using non-80 class counts.

What changed and why it may resolve this issue
- Adds an mlprogram flag to IOSDetectModel to handle MLProgram-specific behavior during CoreML export.
- Pads class scores to the next multiple of 80 when NMS is enabled for MLProgram exports, avoiding a known CoreML NMS bug on iOS.
- Automatically aligns class name mappings when the exported model’s class count doesn’t match provided names, preventing label-related runtime errors.
- Simplifies weights handling by using ct_model.weights_dir directly, reducing export edge cases.

Credit
- Implemented by Y-T-G with contributions from Glenn Jocher.

How to test the fix
- Upgrade to the release with the fix:
  - pip: `pip install -U ultralytics>=8.3.207`
  - or install from source: `pip install git+https://github.com/ultralytics/ultralytics.git@main`
- For detection models, export with CoreML NMS enabled:
  ```bash
  yolo export model=yolo11n.pt format=coreml nms=True
  ```
- If preferred, waiting for the next official PyPI release is also an option.

Request for feedback
- Please try the updated export flow and share whether this resolves the CoreML NMS issue on your iOS setup, particularly with non-80 class models.

Thanks for reporting this and helping improve the export pipeline. If the issue persists, any logs, model details, and reproduction steps would be appreciated.
